File: trunk/matlab/trees/RandomForest.m
classdef RandomForest < handle
    properties
        % trees in the forest
        trees;
        % nmber of trees in the forest
        n_trees;
        % measure of the information gain (gini, entropy)
        gain_measure;
    end
    
    methods
        
        function self = RandomForest(n_trees)
            self.n_trees = n_trees; % number of trees in the forest 
            self.gain_measure = InfoMeasure.gini;
            self.trees = [];
            
        end

        function self = set.gain_measure(self, m)
            self.gain_measure = m;
        end

        
        function self = train(self, data, classes, are_categorical)
            %{
                train a number self.n_trees of decision trees.
                Each tree is trained choosing a random number of datapoints
            
                data - set of features used for training. Each row is a 
                    data point
                classes - the class for each datapoint
                are_categorical - a vector with true if the feature is
                    categorical and false otherwise. Eg [true false true]
                    means that a datapoint has 2 categorical variables, the
                    fist and the third
            %}
            n_points = size(data,1);
            for i=1:self.n_trees
                % Choose random set of data points
                n_choices = ceil(rand(1) * n_points);
                choices = randsample(n_points);
            
                T = DecisionTree;
                T.gain_measure = self.gain_measure;
                T.train(data(choices,:), classes(choices,:), are_categorical);               
                self.trees = [self.trees T];
            end
        end
        
        function classes = predict(self, data)
            % predicts the class of the datapoints in data
            predictions = zeros(size(data,1), self.n_trees);
            for i=1:self.n_trees
                predictions(:,i) = self.trees(i).predict(data);
            end
            % predit based on the most frequent value
            classes = mode(predictions,2);
        end
    end 
    
    
end

File: trunk/cplusplus/src/trees/RandomForest.cpp

#include "trees/RandomForest.h"
#include "utility/eigen_helper.h"
#include "algorithms/sample.h"
#include "utility/errors.h"
#include <stdexcept>
#include <random>


RandomForest::RandomForest(unsigned int n_trees) {
  n_trees_ = n_trees;
  set_information_measure(GINI);
}

void RandomForest::do_training(const MatrixXd &data,
            const VectorXi &classes,
            VariableTypes is_categorical) {
  Ints choices = get_random_sample(data.rows());
  for (unsigned int i = 0; i < n_trees_; ++i) {
    DecisionTreePtr ptr(new DecisionTree());
    ptr->set_information_measure(information_measure_);
    Ints choices = get_random_sample(data.rows());
    ptr->train(data, classes, is_categorical);
    trees_.push_back(ptr);
  }
}

VectorXi RandomForest::get_prediction(const MatrixXd &data) const {
  VectorXi classes(data.rows());
  for (std::size_t i = 0; i < data.rows(); ++i) {
    classes[i] = get_prediction_datapoint(data.row(i));
  }
  return classes;
}

int RandomForest::get_prediction_datapoint(const VectorXd &datapoint) const {
  check_empty();
  VectorXi predictions(n_trees_);
  for (unsigned int i = 0; i < trees_.size(); ++i) {
    predictions[i] = trees_[i]->predict_datapoint(datapoint);
  }
  IntPair p = mode(predictions);
  return p.first;
}

void RandomForest::check_empty() const {
  if(trees_.empty()) {
    throw SizeError("There are no trees in the forest");
  }
}
